hicommonwealth/commonwealth
# New Community Cards + Stake Transactions in non-active community

## Link to Issue
Closes: 
- https://github.com/hicommonwealth/commonwealth/issues/6869
- https://github.com/hicommonwealth/commonwealth/issues/6870

## Description of Changes
1. Uses new community cards with, thread counts, member counts, and buy stake button (for stake-enabled communities)
2. Adds stake buying functionality to global community explore and community directory pages. The stake-buying functionality works slightly differently depending on the context.
    1. From explore page: In this case, we don’t have an `app.activeChainId()` and `app.user.activeAccount`, so when a user finds a stake-enabled community and clicks `Buy Stake`, we filter all their unique addresses and set the first one as selected in the stake transaction modal. If the user has multiple addresses connected, they can select a different address from the dropdown. Any transaction will happen on the selected address. Users don’t need to join the community to buy stakes.
    2. From the community directory page: In this case, we do have an `app.activeChainId()` (id of the current community in which directory is shown, ex: dydx) and optionally `app.user.activeAccount` (if the user is logged into that community). When a user finds a stake-enabled community from the directory page and clicks `Buy Stake`, we filter all their unique addresses (ignoring the user address logged into (if any) currently active community, ex: dydx) and set the first address as selected in the stake transaction modal. If the user has multiple addresses connected, they can select a different address from the dropdown. Any transaction will happen on the selected address. Any transaction will not affect the user’s active address in the current community (ex: if the stake was bought with address-1 in `test-community`, it won’t change the active address in the current community, ex: dydx). Users don’t need to join the community to buy stakes.
    3. From an active community: In this case, we have an `app.activeChainId()` (id of the current community in which directory is shown, ex: dydx) and `app.user.activeAccount` (if the user is logged into that community). 
When the user clicks on the buy stake button from the community sidebar, they are asked to join the community. The address used to join the community is set as active and shown in the stake modal. Any transaction will happen on the selected address. Any transaction with any address will set that address as the active address in that community. The user needs to join the community to buy stakes (this behavior will be changed in https://github.com/hicommonwealth/commonwealth/issues/6587 which would allow user to buy stakes in the active community without membership in that community.).

One logical issue with these points is, if the user bought stake in a community, ex: `test-sepolia` from the global explore page or any community directory page, without joining the community (`test-sepolia`) and then they visit the community (`test-sepolia`), it will show 0 vote weight, this is because there are still logical checks in place for only displaying stake value if user is joined to the active community, this is to be addressed in https://github.com/hicommonwealth/commonwealth/issues/6587

<img width="290" alt="image" src="https://github.com/hicommonwealth/commonwealth/assets/51641047/0fd9d02f-f8dc-4382-bb28-8ce84b6d9e79">


## Test Plan
Pre-req: 
- Add `Sepolia` entry to chain nodes
- Buy some sepolia https://www.alchemy.com/faucets/ethereum-sepolia
- Change wallet network to `Sepolia` and login with any address
- Connect multiple addresses to any community.

1. Visit the global explore page, and any community directory page (to enable the directory, make yourself global admin, then visit Community > Admin Capabilities > Integrations > Enable directory page > select the base chain for directory (for testing use `Sepolia`)), verify that community cards design matches that in [figma](https://www.figma.com/file/omZZrndm6Phbk1rOU4IKhS/Stake-V1-Fast-Follows?type=design&mode=design&t=DkQY7hY46jMASDad-0)
2. Verify that you can buy stakes from the community card in the explore and community directory page, this won't require you to join that community.
3. Verify that you can buy stakes from the active community (i.e being in a community page that has stake enabled), verify that after buying stake, active address changes to the one used to buy stake in that community.
4. Verify that 'Buy stake' is disabled when user is logged out
<img width="475" alt="image" src="https://github.com/hicommonwealth/commonwealth/assets/51641047/2b1e2a22-2d09-4124-973c-df4c38fe1e0c">


## Other Considerations

~~upon joining a new community, the communities listed in the explore communities sidebar is not updated. A bug ticket has been created here: #7099~~ Joining community from non-community pages has been descoped for now https://github.com/hicommonwealth/commonwealth/issues/6872#issuecomment-1979218770
